sw/host/hsmtool/src/util/key/rsa.rs

```diff
@@ -0,0 +1,366 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{Context, Result};
+use rsa::pkcs1::{
+    self, DecodeRsaPrivateKey, DecodeRsaPublicKey, EncodeRsaPrivateKey, EncodeRsaPublicKey,
+};
+use rsa::pkcs8::{self, DecodePrivateKey, DecodePublicKey, EncodePrivateKey, EncodePublicKey};
+use rsa::{PublicKeyParts, RsaPrivateKey, RsaPublicKey};
+use std::convert::{AsRef, TryFrom};
+use std::path::Path;
+
+use super::KeyEncoding;
+use crate::error::HsmError;
+use crate::util::attribute::{AttrData, AttributeMap, AttributeType, KeyType, ObjectClass};
+use crate::util::helper;
+
+fn _load_private_key(path: &Path) -> Result<RsaPrivateKey> {
```

Comment:
This is a common trick to prevent bloat from monomorphization.  The private versions of the functions accept `&Path`, while the public versions accept any type `P where P: AsRef<Path>` (which includes, Path, PathBuf, str, String, OsStr, OsString).

Rather than allowing monomorphization to emit separate versions of a "large" function for every observed input type, the trick is to separate out the common code into a private function with a non-generic signature and then have the public function perform whatever parameter conversion is needed (e.g. `P::as_ref()`) and then call the private function.